### PR TITLE
Onboarding: Re-order pages to hide loading behind notifications prompt

### LINKED ISSUE #186: Onboarding: Re-order pages and deploy in background
Besides #164, one thing we can do is re-order onboarding screens to make the flow such that Create Account/Add Existing pages are shown first, then we prompt for notifications. In background, we can start the deploy already during the user clicking notifications prompts etc. (and if they click through fast enough we can show the loader as right now).

Would make it feel fast. It also feels like notifications should probably come after the first step (since they're optional as well and less important than create account/add existing).

### PR DESCRIPTION
**Pull request by nalinbhardwaj**
Closes #186 

Based on #201 don't merge till that's merged.

### CODE REVIEW

File: apps/daimo-mobile/src/view/HomeStack.tsx [hunk 1]
```diff
@@ -67,6 +67,13 @@ export function HomeStackNav() {
 
 function MainScreen() {
   const [account] = useAccount();
-  if (account) return <HomeScreen />;
-  return <OnboardingScreen />;
+  const [isOnboarding, setIsOnboarding] = useState<boolean>(false);
+  const onOnboardingComplete = () => setIsOnboarding(false);
+
+  useEffect(() => {
+    if (!account) setIsOnboarding(true);
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/HomeStack.tsx:74]: you can say

```
const [isOnboarding, setIsOnboarding] = useState(account == null);
```

File: apps/daimo-mobile/src/view/screen/OnboardingScreen.tsx [hunk 2]
```diff
@@ -36,35 +38,93 @@ import {
...
+          }
+        })()
+      );
     },
     [page]
   );
 
-  // TODO: add back buttons?
+  // TODO: add back buttons on create and existing pages
+  const prev = useCallback(() => {}, []);
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/OnboardingScreen.tsx:83]: easy to implement now that we've refactored `page`

copy above, edit switch cases ^

File: packages/contract/src/Account.sol [hunk 3]
```diff
@@ -55,29 +68,13 @@ contract Account is BaseAccount, UUPSUpgradeable, Initializable {
...
-        address[] calldata dest,
-        bytes[] calldata func
-    ) external {
+    function executeBatch(Call[] calldata calls) external {
         _requireFromEntryPoint();
-        require(dest.length == func.length, "wrong array lengths");
-        for (uint256 i = 0; i < dest.length; i++) {
-            _call(dest[i], 0, func[i]);
+        for (uint256 i = 0; i < calls.length; i++) {
+            _call(calls[i].dest, calls[i].value, calls[i].data);
```
Comment (dcposch) [inline packages/contract/src/Account.sol:77]: 🚀 nice cleanup

File: packages/contract/src/Account.sol [hunk 4]
```diff
@@ -109,7 +117,10 @@ contract Account is BaseAccount, UUPSUpgradeable, Initializable {
             userOpHash
         ); // Emulate EIP-712 signing: https://eips.ethereum.org/EIPS/eip-712
         uint8 keySlot = uint8(userOp.signature[0]);
-        require(keys[keySlot][0] != bytes32(0) && keys[keySlot][1] != bytes32(0), "invalid key slot");
+        require(
+            keys[keySlot][0] != bytes32(0) && keys[keySlot][1] != bytes32(0),
+            "invalid key slot"
```
Comment (dcposch) [inline packages/contract/src/Account.sol:122]: nit but you could check just `keys[keySlot][0]` here & below

File: packages/contract/src/Account.sol [hunk 5]
```diff
@@ -186,7 +204,10 @@ contract Account is BaseAccount, UUPSUpgradeable, Initializable {
      * @param key the P256 public key to add
      */
     function addSigningKey(uint8 slot, bytes32[2] memory key) public onlySelf {
-        require(keys[slot][0] == bytes32(0) && keys[slot][1] == bytes32(0), "key already exists");
+        require(
+            keys[slot][0] == bytes32(0) && keys[slot][1] == bytes32(0),
+            "key already exists"
+        );
```
Comment (dcposch) [inline packages/contract/src/Account.sol:210]:
```solidity
require(keys[slot][0] == bytes32(0), "key already exists");
require(key[0] != bytes32(0), "new key cannot be 0");
```

File: packages/contract/src/AccountFactory.sol [hunk 6]
```diff
@@ -26,19 +28,25 @@ contract AccountFactory {
...
-        address addr = getAddress(accountKey, salt);
+    ) public payable returns (Account ret) {
+        address addr = getAddress(slot, key, initCalls, salt);
         uint codeSize = addr.code.length;
         if (codeSize > 0) {
             return Account(payable(addr));
         }
+
+        // Prefund the account with msg.value
+        entryPoint.depositTo{value: msg.value}(addr);
```
Comment (dcposch) [inline packages/contract/src/AccountFactory.sol:43]: might wrap this in `if (msg.value > 0)`

saves us a contract call when using paymaster

File: packages/contract/test/SigningKeys.t.sol [hunk 7]
```diff
@@ -22,8 +22,16 @@ contract SigningKeysTest is Test {
...
+    event SigningKeyAdded(
+        IAccount indexed account,
+        uint8 keySlot,
+        bytes32[2] key
+    );
+    event SigningKeyRemoved(
+        IAccount indexed account,
+        uint8 keySlot,
+        bytes32[2] key
+    );
```
Comment (dcposch) [inline packages/contract/test/SigningKeys.t.sol:34]: 👍 👍

File: packages/daimo-api/src/router.ts [hunk 8]
```diff
@@ -289,32 +299,39 @@ export function createRouter(
...
+            data: encodeFunctionData({
+              abi: erc20ABI,
+              functionName: "approve",
+              args: [ephemeralNotesAddress, maxUint256],
+            }),
+          },
+          nameReg.getRegisterNameCall(name), // Register name
+        ];
+
+        // TODO: put a check for the counterfactual address on client side so the server is not trusted.
```
Comment (dcposch) [inline packages/daimo-api/src/router.ts:317]: 👍

**Review by dcposch — COMMENTED** [review]
Looks great to me, left comments

File: apps/daimo-mobile/src/view/screen/OnboardingScreen.tsx (hunk 2 above)
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/screen/OnboardingScreen.tsx:83]: will leave for later, requires implementing a "header" and intercepting android back buttons as well.

File: packages/contract/src/Account.sol (hunk 4 above)
Comment (nalinbhardwaj) [reply packages/contract/src/Account.sol:122]: done

File: packages/contract/src/AccountFactory.sol (hunk 6 above)
Comment (nalinbhardwaj) [reply packages/contract/src/AccountFactory.sol:43]: done

File: packages/contract/src/Account.sol (hunk 5 above)
Comment (nalinbhardwaj) [reply packages/contract/src/Account.sol:210]: done

**Pull request merged**